inets: Remove default empty TE header from HTTP requests

iamyojimbo opened the pull request:
The httpc client was unconditionally setting the TE header to an empty string on all HTTP/1.1 requests when no TE header was explicitly provided. This behavior violated RFC 2616 and caused compatibility issues with some proxy servers that would drop requests containing empty TE headers.

RFC 2616 does not require a TE header to be present, and when absent, no transfer encoding expectations should be assumed. This change removes the automatic addition of empty TE headers, allowing httpc to send cleaner HTTP requests that better comply with the standard.

Fixes GH-10065

Whaileee commented:
Thanks for the contribution. Change looks good to me. Please replace the strip with trim, and I think it's good to go

Whaileee commented:
Hi, there seems to be an error in all of the added testcases, when run in group sim_http_ipv6. All tests fail with connection header "close".
Did you run ipv6 tests? Are you able to reproduce that?
Best regards,
Konrad

iamyojimbo commented:
@Whaileee - please take another look - tests passing on my side. I changed ipv6 to not send per-request connection options (`socket_opts`), because that cases httpc to override all request connection headers with `connection: close`. Looks like there's some emotional baggage around that in the code: https://github.com/erlang/otp/blob/286418c84cba4946b781891fa9ad560a2f91d94c/lib/inets/src/http_client/httpc_manager.erl#L743-L760

Rather than change `httpc_manager: handle_request` to preserve tokens of the existing connection header, I just split out  `sim_http_ipv6` group to use `httpc_options` instead of `socket_opts` and left `http_ipv6` as-is.

Let me know if these test pass now.

Whaileee commented on lib/inets/test/httpc_SUITE.erl:
```diff
@@ -1689,6 +1706,62 @@ redirect_port_in_host_header(Config) when is_list(Config) ->
+%%-------------------------------------------------------------------------
+te_header_undefined_no_connection(Config) when is_list(Config) ->
+    Request = {url(group_name(Config), "/te_header_undefined_no_connection.html", Config), []},
+    RequestOpts = proplists:get_value(request_opts, Config, []),
```
After splitting the init_per_group for http_ipv6 and sim_http_ipv6, aren't the RequestOpts always an empty list?
Also the request is now sent using ipv4, or am I missing something?

iamyojimbo replied:
True - although, if we do remove this line, if someone adds `request_opts` in `init_per_group` at any point, then these tests would not fail, when they should have.

We could change to something like:

```
?assertEqual("request_opts must be empty until httpc_manager:handle_request/2 is fixed to handle per-request opts with persistent connections",
  [],
  proplists:get_value(request_opts, Config, [])).
```

Thoughts?

iamyojimbo commented:
@Whaileee How'd the tests go?

iamyojimbo commented:
@Whaileee Any luck?

Whaileee commented:
For some reason the branch was still commented out of testing. I've now added it manually, so if it's not in testing now I'll start breaking stuff (just kidding). Therefore I ask for another day of patience.

Whaileee commented:
We will see. The tests on the machines with ipv6 are still running. Should be done in a few hours

Whaileee commented:
That is true for 26 and 27, but for 28 it will be in 28.1

iamyojimbo commented:
🙏

iamyojimbo commented:
Just to clarify: by minor patch, do we mean quarterly minor patches?

Whaileee commented:
This branch will be released with OTP 27.3.4.3 and OTP 26.2.5.15 which are planned to be released soon, as well as OTP 28.1 which will be also released soon-ish.
